File: python/ProdVectMat-Gather.py
#!usr/bin/python3


from mpi4py import MPI
import numpy as np
import time
import sys

# Parallel vector-matrix multiplication using MPI

# Initializzing Parallel part of the code
comm = MPI.COMM_WORLD # global communicator
rank = comm.Get_rank() # extracting task id
size = comm.Get_size() # extracting number of processes

# starting time measuring
t0 = time.time()

# Read the size of the matrix from a file
with open('input_n.txt') as file:
    n = int(file.read().strip()) # removing blank characters


# and initialize the vector V, values from 1 to n
V = np.array(np.arange(1,n+1), dtype='f8')

# defining number of columns to assing to the current task (to be adjusted later)
ncolumn = n//size
remainder = n%size

# initializing the arrays for GatherV communication
recvcounts=[] # size of the block of each task
displs = []   # starting index of each block

disp=0
# computing recvcounts and displs arrays
for irank in range(size):
    displs.append(disp)
    if irank < remainder:
        incol = ncolumn + 1
    else:
        incol = ncolumn
    recvcounts.append(incol)
    disp += incol

# adjusting size of each block of the matrix
if rank < remainder:
    ncolumn += 1


# initialization of portion of matrix M of ncolumns columns for each task
# (elements' value of the whole matrix ranging from 1 to n for each row)
M = np.array([np.arange( displs[rank] +1, 
                           displs[rank] +1 + recvcounts[rank]) 
                           for i in range(n)], dtype='f8') 


# computing the multiplication on the task's portion of matrix
C = V@M


## MPI communication for retrieving the resulting vector
## using GatherV function

buffer = np.zeros(n) # receive buffer
# gather to task 0 all the local C result vectors
comm.Gatherv(C, recvbuf=(buffer, np.array(recvcounts), np.array(displs), MPI.DOUBLE), root=0)

## saving the results (first 100 elements to a file)

name = sys.argv[0].split('/')[-1].split('.')[0] # defining output file
if rank==0:
    ## output filename= "<program name>_py_output.txt"
    with open(name + '_py_output.txt', 'w') as f:
        np.savetxt(f, buffer[:100], fmt='%.6f')
    #print(buffer[495:505]) # prints for consistency check of the results
    #print(buffer[-5:])

t_tot = time.time() - t0

## saving computation time to a file for computing performance metrics
if rank == 0:
    ## filename= "times_< matrix size >_<program name>_py.txt"
    timefile = f'times_{n}_' + name + '_py.txt' 
    with open(timefile, 'a') as f:
        f.write(f'{size} {t_tot}\n')

File: fortran/ProdVectMat-Gather.f90
program vect_mat_mul
    USE MPI
    implicit none
    
    integer, parameter :: K2 = selected_real_kind(15, 30)
    integer :: n,i,j,k,st, remainder, disp
    integer :: size, ierr, ecode, rank , status(MPI_STATUS_SIZE)
    real :: cpu1, cpu2, deltat
    real(KIND=K2), allocatable, dimension(:,:) :: M
    real(KIND=K2), allocatable, dimension(:) :: V, C, buffer
    integer, allocatable, dimension(:) :: DISPLS, RECVCOUNTS
    integer :: Ncolumns
    character(len=100) :: times_filename

    call MPI_INIT(ierr)
    call MPI_COMM_SIZE(MPI_COMM_WORLD,size,ierr)
    call MPI_COMM_RANK(MPI_COMM_WORLD,rank,ierr)
    
    CALL CPU_TIME(cpu1)

    print*, 'MPI initialized. Rank:', rank, 'of', size

    OPEN(11,FILE="input_n.txt",STATUS="OLD",IOSTAT=st)
    READ(11,*,IOSTAT=st) n
    CLOSE(11, STATUS='KEEP')

    print*, 'Dimension of the matrix:', n

    !computing number of columns that each process will handle
    print*, 'Number of processes:', size
    Ncolumns = n / size
    remainder = mod(n, size)
    print*, 'Number of columns per process:', Ncolumns

    ! Setting DISPLS and RECVCOUNTS for MPI_GATHER
    allocate(DISPLS(size))
    allocate(RECVCOUNTS(size))

    disp = 0
    do i=0, size-1
        DISPLS(i+1) = disp
        if (i < remainder) then
            RECVCOUNTS(i+1) = Ncolumns + 1
        else
            RECVCOUNTS(i+1) = Ncolumns
        end if
        disp = disp + RECVCOUNTS(i+1)
    end do

    print*, 'DISPLS:', DISPLS
    print*, 'RECVCOUNTS:', RECVCOUNTS
    !if 

    ! If the number of columns is not divisible by size increase by one
    ! the first elements
   
    if (rank < remainder) then
        Ncolumns = Ncolumns + 1
    end if

    print*, 'Adjusted number of columns for process', rank, ':', Ncolumns

    ! Initialize Vector
    allocate(V(n))   
    
    do i=1,n
        V(i) = i
    end do

    ! Initialize Matrix
    allocate(M(n,Ncolumns))

    do i=1,n
        do j = 1, Ncolumns
        !do j=DISPLS(rank)+1, DISPLS(rank)+1 + RECVCOUNTS(rank)
            M(i, j) = DISPLS(rank+1) + j
        end do
    end do
    
    ! do i=1,n
    !     print*, M(i,:)
    ! end do

    ! initialize resulting vector
    allocate(C(n))

    C = MATMUL(V, M)

    !communication part
    allocate(buffer(n))
    call MPI_GATHERV(C, Ncolumns, MPI_DOUBLE_PRECISION, buffer, RECVCOUNTS, DISPLS, MPI_DOUBLE_PRECISION, 0, MPI_COMM_WORLD, ierr)
    if (rank == 0) then
        ! The root process gathers the results
        ! print '(A,10(E15.7E3,1X))', 'Middle ten gathered results:', buffer(496:505)
        ! print*, 'Last five elements of the gathered vector:', buffer(n-4:n)
        OPEN(12,FILE="ProdVectMat-Gather_f90.output",STATUS="REPLACE")
        do i=1, 100
            WRITE(12,*) buffer(i)
        end do
        CLOSE(12, STATUS='KEEP')
    end if

    CALL CPU_TIME(cpu2)

    deltat = cpu2 - cpu1
    if (rank == 0) then
        WRITE(times_filename, '(A,I0,A)') 'times_', n, '_ProdVectMat-Gather_f90.txt'
        OPEN(13,FILE=times_filename,status='unknown', position='append')
        WRITE(13, '(I2, X, F10.6)') size, deltat
        CLOSE(13, STATUS='KEEP')
        !print*, 'Time taken for initialization:', deltat
    end if

    CALL MPI_FINALIZE(ierr)
end program vect_mat_mul

